New document_exporter options

mhelleboid opened the pull request:
## Proposed change

Added few additional options to management document_exporter and a bug fix.
Everything in the same PR but in individual commits to avoid to many PR dependencies.

- Added exporter option to exclude archive and thumbnails in export directory: these files can be recreated with document_archiver or document_thumbnails
- Added exporter option to generate json document information in dedicated json file (each document has an associated json file)
- Added exporter option to prefix files with their document nature : archive, thumbnails, original, json
- Fix a bug with the -d --delete exporter option (version.json file and dedicated json files were deleted)

I can also open an issue for the fixed bug, tell me if this is needed

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

mhelleboid committed:
add no-archive and no-thumbnail options to administration exporter and importer

mhelleboid committed:
add prefix option to administration exporter

mhelleboid committed:
add split-manifest option to administration exporter

mhelleboid committed:
fix bug on administration exporter when using -d, --delete option

mhelleboid committed:
fix json serialization bug after migration after to Pathlib)

shamoon commented:
Thanks, we will look at some point. FYI you dont need to recreate the PR every time eg. you need to resolve conflicts, you can (force-) push to this branch and itll update.

mhelleboid commented:
Yes, sorry for that, this was not intentional, I closed the first one using the UI by mistake, and the second one when I forced push my fork/dev branch.

stumpylog commented:
The bugfix is good, thank you for that, it's something I missed when implementing the export versioning.

But before I get too much into reviewing the code, I'd like to understand the use case for some of these options.  I don't want to bring in complexity to such an important tool without good reason.

- What is the benefit of multiple JSON files over a single JSON manifest?
- What does prefixing files with a fixed string help a user do or understand?
- Although archive and thumbnails can be generated again, it requires the user to do so manually and the sanity checker will  be mad until they do.  How is making a user do extra work beneficial?

mhelleboid commented:
I understand your point, here is my reasons:

> * What is the benefit of multiple JSON files over a single JSON manifest?
For a usecase where this export is used as backup (or in git).
If a new document is added/removed to paperless, it makes it much easier to know what has changed to add/remove a smaller json file than modifying a big manifest file.

> * What does prefixing files with a fixed string help a user do or understand?
In my situation, backups are based on this export. And I only want to backup the original files. I makes it much easier to choose what to backup or not when these files are splitted in different root folders

> * Although archive and thumbnails can be generated again, it requires the user to do so manually and the sanity checker will  be mad until they do.  How is making a user do extra work beneficial?
This is just an option and is optional for users.
In my situation I never want to export thumbnails (so the --no-thumbnails option)
The --no-archive option was easy to implement and could also be useful for users who do not want also to backup archive files without modifying the backup program. It was also useful to test the reimport without these archive files.

stumpylog commented on docs/administration.md:
```diff
@@ -259,6 +266,14 @@ current export such as files from deleted documents, specify `--delete`.
 other files.
 
+Paperless will not export archive files if you use `--no-archive`, or will
+not export thumbnails if you use `--no-thumbnail`. After importing, These
```
For both the no thumbnail and no archive options, there should be a clear warning that the sanity checker will warn about missing files and the documents view will not have thumbnails until the user intervenes manually

mhelleboid replied:
fix by c0466755d8b3cc0ae5dbed6993b4392d83f210c6

stumpylog replied:
I tweaked how the note was displayed

stumpylog commented on src/documents/management/commands/document_exporter.py:
```diff
@@ -243,15 +285,24 @@ def dump(self, progress_bar_disable=False):
             # 3.3. write filenames into manifest
             original_name = base_name
+            if self.use_filename_prefix:
+                original_name = os.path.join("originals", original_name)
```
Lets keep to using `Path`s where possible.

mhelleboid replied:
Try that at first, but ended up trying to store Pathlib in the json then, with an exception
see e03c72a584fd7810257f400729c63505d76f20a3

stumpylog commented on src/documents/management/commands/document_importer.py:
```diff
@@ -72,11 +72,24 @@ def handle(self, *args, **options):
+
+        for root, dirs, files in os.walk(self.source):
+            for file in files:
+                if file.endswith("-manifest.json"):
```
Pathlib globbing should be a quicker way to locate `*.json` and perhaps even `*-manifest.json`.  It would be something like `.glob(**/*-manifest.json)`

mhelleboid replied:
fix by ead820eeb62b33872a3f4eaa63a6cafc7373e4dc

stumpylog commented on src/documents/tests/test_management_exporter.py:
```diff
@@ -497,3 +506,80 @@ def test_export_target_not_writable(self):
 
                 self.assertEqual("That path doesn't appear to be writable", str(e))
+
+    def test_no_archive(self):
```
It's great to include testing.  Please use the [given/when/then](https://en.wikipedia.org/wiki/Given-When-Then) style commenting to help quickly show what the test is doing and the expected output

mhelleboid replied:
fixed by c0466755d8b3cc0ae5dbed6993b4392d83f210c6

stumpylog commented on docs/administration.md:
```diff
@@ -227,19 +227,26 @@ is not a TTY" errors. For example:
+document_exporter target [-sm] [-c] [-f] [-p] [-d] [-na] [-nt]
 
 optional arguments:
+-sm, --split-manifest
```
I don't think there was a plan with these before, but I think it would be nice to see the options sorted alphabetically, by the short name

mhelleboid replied:
fixed by 202d8a70884d76d0abf5ea37e65dba1b20b78f18

stumpylog commented on docs/administration.md:
```diff
@@ -227,19 +227,26 @@ is not a TTY" errors. For example:
+-sm, --split-manifest
 -c, --compare-checksums
 -f, --use-filename-format
+-p, --use-filename-prefix
```
I feel like this could be better named.  It separates the various files into directories, which is more than just a change to a filename.

stumpylog replied:
Perhaps `--use-folder-prefix` since it's about folders more than filenames?

mhelleboid replied:
Agreed, fixed by 7eb784d08e9bdf84f1ab6b9b41299dee23b84f80

mhelleboid committed:
Fix split_manifest default value

mhelleboid committed:
fix long option name for "-p" or "--use-filename-prefix"

mhelleboid committed:
--no-archive & --no-thumbnails option has their own paragraph

mhelleboid committed:
add warning about --no-archive & --no-thumbnails options
add given_when_then informaiton to tests
add use_filename_prefix test
clear documents before importing in tests

mhelleboid committed:
add document comments to dedicated manifest file when using 'split-manifest'

mhelleboid committed:
use pathlib glob

mhelleboid committed:
sort exporter option by alphabetical order

mhelleboid commented on src/documents/management/commands/document_exporter.py:
```diff
@@ -100,18 +133,26 @@ def add_arguments(self, parser):
     def __init__(self, *args, **kwargs):
         BaseCommand.__init__(self, *args, **kwargs)
         self.target: Path = None
+        self.split_manifest = None
```
fixed by 8b5fb12f492109ed66e12f3aab8851dafdbcc6d0

mhelleboid commented on docs/administration.md:
```diff
@@ -267,6 +282,10 @@ The filenames generated by this command follow the format
 paperless to use `PAPERLESS_FILENAME_FORMAT` for exported filenames
 instead, specify `--use-filename-format`.
 
+If `-p` or `--use-filename-format` is provided, Files will be exported
```
fix by 42216b1ba57e12ab4b5fa98c4e316b499efabf40

mhelleboid commented on docs/administration.md:
```diff
@@ -259,6 +266,14 @@ current export such as files from deleted documents, specify `--delete`.
 Be careful when pointing paperless to a directory that already contains
 other files.
 
+Paperless will not export archive files if you use `--no-archive`, or will
```
fix by 2c400c45386ec20ac092728ba1b98ef1cd64c017

mhelleboid commented:
@stumpylog, thanks for the very detailed review. 
I made modifications according to your comments, I suppose I must let you resolve the conversations.
Also fixed a bug with "--split-manifest" and comments in 5e5b1da265abc8e1e73244840b32fbd45d5cddac

stumpylog commented:
The changes look pretty good from a quick glance.  I'll have more time this weekend to complete reviewing.

stumpylog committed:
Tweaks the wording and note about usage of the -no options

stumpylog commented on src/documents/management/commands/document_exporter.py:
```diff
@@ -292,22 +347,40 @@ def dump(self, progress_bar_disable=False):
+
         # 4.1 write manifest to target folder
         manifest_path = (self.target / Path("manifest.json")).resolve()
         manifest_path.write_text(json.dumps(manifest, indent=2))
```
When I set `--split-manifest` the original `manifest.json` is also created still.  Probably needs a check around here?

mhelleboid replied:
Yes, this is expected. Information about documents are in dedicated json files, but application settings are still in the main manifest (tags, correspondent, documenttype, etc)

stumpylog replied:
Ah, I see.  Thanks

mhelleboid replied:
But maybe, if you wondered why, others will too. And then we should explain that in the documentation

mhelleboid replied:
see 16f008d5c6d12b0114a2f07967d7d8e7eeb3d52b

mhelleboid committed:
replace --use-filename-prefix with --use-folder-prefix

mhelleboid committed:
add information about main manifest for -sm
